### PR TITLE
feat: first step of `openmp` pass

### PR DESCRIPTION
**Pull request by Pranavchiku**
Towards #4137, #3777.

With current status of pass we transform

```fortran
subroutine omp_func(n)
implicit none
integer, intent(in) :: n
integer :: i

!$omp parallel private(i) shared(n)
!$omp do
do i = 1, n
    print *, "xyz"
end do
!$omp end do
!$omp end parallel

print *, "n = ", n
end subroutine

program openmp
integer, parameter :: n = 100
call omp_func(n)
print *, "Done for n = ", n
end program
```

TO

```fortran
module thread_data_module
    use, intrinsic :: iso_c_binding
    type, bind(C) :: thread_data
        integer(c_int) :: n
    end type thread_data
end module thread_data_module

subroutine omp_func(n)
use iso_c_binding
use thread_data_module
implicit none
integer(c_int), intent(in) :: n
integer :: i

type(thread_data), target :: data
type(c_ptr) :: tdata

data%n = n
tdata = c_loc(data)

print *, "n = ", n
end subroutine

program openmp
use iso_c_binding
integer(c_int), parameter :: n = 100
call omp_func(n)
print *, "Done for n = ", n
end program
```

### CODE REVIEW

File: integration_tests/openmp_01.f90 [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+!$omp parallel private(i) shared(n)
+!$omp do
+do i = 1, n
+    ! print *, "xyz"
+end do
+!$omp end do
+!$omp end parallel
+
+print *, "n = ", n
+if (n /= 100) error stop
```
Comment (Pranavchiku) [inline integration_tests/openmp_01.f90:15]: This is just a dummy test which we aim to compile, right now it is not transformed to desired output but an intermediate step for which `if ( n/= 100 ) error stop` will work.

File: src/libasr/pass/replace_openmp.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+#include <libasr/utils.h>
+
+namespace LCompilers {
+
+    void pass_replace_openmp(Allocator &al, ASR::TranslationUnit_t &unit,
+                                const PassOptions &pass_options);
+
+} // namespace LCompilers
+
+#endif // LIBASR_PASS_REPLACE_OPENMP
```
Comment (certik) [inline src/libasr/pass/replace_openmp.h:14]: Register this pass in `src/libasr/gen_pass.py` and let it generate this file (but still check it in).

File: src/bin/lfortran.cpp [hunk 3]
```diff
@@ -2143,6 +2143,8 @@ int main_app(int argc, char *argv[]) {
     CLI11_PARSE(app, argc, argv);
     lcompilers_unique_ID = compiler_options.generate_object_code ? get_unique_ID() : "";
 
+    compiler_options.po.openmp = compiler_options.openmp;
```
Comment (certik) [inline src/bin/lfortran.cpp:2146]: What is this for? Let's add a comment to it, and perhaps move it further down where other `compiler_options` are processed.

File: src/libasr/pass/openmp.cpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                        m->m_name, nullptr, 0, temp->m_name,
+                        ASR::accessType::Public
+                    );
+                    current_scope->add_symbol(item.first, ASR::down_cast<ASR::symbol_t>(v));
+                } else {
+                    return item.first;
+                }
+            }
+            return "";
+        }
```
Comment (certik) [inline src/libasr/pass/openmp.cpp:190]: This functions should probably be refactored into `asr_utils.h`. This can be done later.

**Review by certik — APPROVED** [review]
I think that looks good, thanks! I left some comments.

File: src/bin/lfortran.cpp (hunk 3 above)
Comment (Pranavchiku) [reply src/bin/lfortran.cpp:2146]: We need to have openmp to pass options so I did it here. I’ll move it further down.

File: src/libasr/pass/openmp.cpp (hunk 4 above)
Comment (Pranavchiku) [reply src/libasr/pass/openmp.cpp:190]: Yes, that was my initial thought as it is also used in `ast_symboltable_visitor`. I kept it there to fully populate it and see if any changes are required, then I'll shift it to `asr_utils.h`

**Pull request merged**